### PR TITLE
Add ADR for Plugin Package Structure

### PR DESCRIPTION
**Pull request by Fox32**
As we have the demand for more plugin packages right now, here is an ADR that defines a schema.

This ADR is explicitly a RFC and just a first draft.

Things to think about:
* Initially @freben suggested to only reexport the `x-common` package, but I think that requires to always have a `x-common-node` or `x-common-react` which might be some overhead.
* Do we really want to have an exception for `catalog-model`?

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Fox32) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:46]: It is kind of strange that the package name isn't the same as the directory name (but they are all in the `plugins` directory, so I think it should be fine)

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Fox32) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:68]: Would also suggest to extend `create-plugin` with flags for the package types.

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (adamdmharvey) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:22]:
```suggestion
> There is a common phrase in software development:
```

maybe this?

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (adamdmharvey) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:49]:
```suggestion
This structure is based on a [suggestion in issue
#3655](https://github.com/backstage/backstage/issues/3655#issuecomment-758166746).
```

In general, trying to avoid making the word "here" or "click here" clickable is preferred for sight reading (where your eye scans for underscored words or phrases) and accessibility readers (where they're read relevant links, and the word "here" won't make sense in context).

**Review by adamdmharvey — COMMENTED** [review]
Seems to make good sense.

One area not covered is those plugins not a part of the main repo.  Taking a quick look, the marketplace currently has 11 which aren't stored in this repo and their naming doesn't quite match up.  Should these references apply to those?  Or at least a paragraph to call out those and how this ADR intends to, or does not intend to, apply to them.

```
$ grep npmP *.yaml | grep -v "@backstage"
argo-cd.yaml:npmPackageName: '@roadiehq/backstage-plugin-argo-cd'
aws-lambda.yaml:npmPackageName: '@roadiehq/backstage-plugin-aws-lambda'
buildkite.yaml:npmPackageName: '@roadiehq/backstage-plugin-buildkite'
firebase-functions.yaml:npmPackageName: '@roadiehq/backstage-plugin-firebase-functions'
github-insights.yaml:npmPackageName: '@roadiehq/backstage-plugin-github-insights'
github-pull-requests.yaml:npmPackageName: '@roadiehq/backstage-plugin-github-pull-requests'
jira.yaml:npmPackageName: '@roadiehq/backstage-plugin-jira'
opsgenie.yaml:npmPackageName: '@k-phoen/backstage-plugin-opsgenie'
security-insights.yaml:npmPackageName: '@roadiehq/backstage-plugin-security-insights'
snyk-security.yaml:npmPackageName: 'backstage-plugin-snyk'
travis-ci.yaml:npmPackageName: '@roadiehq/backstage-plugin-travis-ci'
```

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (OrkoHunter) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:56]:
```suggestion
- `packages/techdocs-common` which should be moved to
```

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Rugvip) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:59]: Would that be `@backstage/plugin-catalog-common`?

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Rugvip) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:40]: Thoughts about just calling these `x-react` and `x-node`?

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Rugvip) [reply docs/architecture-decisions/adr011-plugin-package-structure.md:68]: There's an issue to switch `@backstage/create-app` to just `@backstage/create`. We could do the same for the CLI, where you use `backstage-cli create` to create ALL the things!

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Fox32) [reply docs/architecture-decisions/adr011-plugin-package-structure.md:40]: They are getting quite long indeed, for me it would be fine to omit the `common-` here.

Comment (Fox32) [reply docs/architecture-decisions/adr011-plugin-package-structure.md:40]: Any other opinions before I change that?

Comment (freben) [reply docs/architecture-decisions/adr011-plugin-package-structure.md:40]: Totally ok. They aren't necessarily related as such (in a hierarchical sense) to common anyway. Sounds good

Comment (Fox32) [reply docs/architecture-decisions/adr011-plugin-package-structure.md]: Updated, looks like I have to rename some stuff in #4265 😆

**Review by freben — APPROVED** [review]
Thanks!

I wonder what the balance will be between `x` and `x-react` in terms of what goes where. In one extreme, `x` would be pretty much completely empty except for the `plugin.ts` file. But I assume we'll often put e.g. Page level components there, and such things that aren't very reusable.

**Comment by Fox32** [comment]
> I wonder what the balance will be between `x` and `x-react` in terms of what goes where. In one extreme, `x` would be pretty much completely empty except for the `plugin.ts` file. But I assume we'll often put e.g. Page level components there, and such things that aren't very reusable.

While working on `catalog-react` I also noticed that most of the code could go into it. I think that `catalog`/`x` should not have any public API except the plugin registration.

**Comment by freben** [comment]
Also: `CatalogApi` in `x-common`, and `catalogApiRef` in `x-react` and the latter reexports the former... Or nah?

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Rugvip) [inline docs/architecture-decisions/adr011-plugin-package-structure.md:38]: Having `react` in the name is both good and bad imo. I'd assume we'll want to stick to that name only for packages that are pretty much exclusively react. Some other options could be `-lib` or `-web`. I think `-react` is good for now but it could be something to keep an eye on

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Rugvip) [reply docs/architecture-decisions/adr011-plugin-package-structure.md]: > Also: `CatalogApi` in `x-common`, and `catalogApiRef` in `x-react` and the latter reexports the former... Or nah? There's a certain level of inconvenience to have to depend on several packages for "no reason" and have several separate import lines etc.
> 
> Is reexporting a bad thing? Opinions?

I think reexport of the API type makes sense in this case, so that you don't have to add a dependency on `-common`. Tbh I think it might be tricky to establish a pattern right now, and look at it case by case. One thing to look out for is whether you can completely remove the need for a package in the app or in plugins by re-exporting some things.

File: docs/architecture-decisions/adr011-plugin-package-structure.md
Comment (Fox32) [reply docs/architecture-decisions/adr011-plugin-package-structure.md:38]: For the backend we could use `x-backend-common` (instead of the `x-common-node` we had first`, but we can't directly use the same pattern in the frontend 😆 But we could still use something like `x-frontend-common`.

On the other hand I think that most frontend code is either react code, or frontend code that is only used in the context of react. Once we come into the situation that we have multi-framework code we would have to refactor some stuff anyway and can touch these rules again.

**Comment by Fox32** [comment]
> Also: `CatalogApi` in `x-common`, and `catalogApiRef` in `x-react` and the latter reexports the former... Or nah? There's a certain level of inconvenience to have to depend on several packages for "no reason" and have several separate import lines etc.
> 
> Is reexporting a bad thing? Opinions?

I would not create packages that are just there to reexport stuff because this document says so, but if you already have `x-react` anyway it sounds like a neat thing. Do we want to explicitly state that `x-react` and `x-node` should reexport `x-common` if they exist?

**Comment by freben** [comment]
Maybe leave out the reexport thing until we learn more. I think it may be a good thing to do, but maybe we can get some experience with this first and then see if we should tweak it a bit. @Rugvip agree good to merge?

**Comment by andrewthauer** [comment]
To @adamdmharvey's point, it might be helpful to also provide some suggested guidance on naming of packages for 3rd party and closed sourced plugins as well. It seems like it has already been established to use `@org/backstage-plugin-` or `backstage-plugin-` as prefixes. We've choose to also name our internal packages using this naming convention for all packages, including the app & backend.l (minus the `plugin` specifier). This would help people find these packages more consistently imo.

**Comment by freben** [comment]
our own create-plugin script gives them no namespace at all per default, which then probably should be the default here as well.

**Comment by andrewthauer** [comment]
Maybe the CLI could ask for an org name as part of the wizard and use that as scope for apps and plugins? It could also add the scope to the root package.json and use that moving forward as the default for creating new plugins.

**Comment by Rugvip** [comment]
@andrewthauer We use `@internal` as the scope internally, and iirc also recommend that in docs. Perhaps defaulting to `@internal` instead of no scope makes sense? Requiring a `--no-scope` or `--scope false` / `--scope ''` to remove it.

We want to avoid forcing those kind of decisions to a wizard btw, since it's something that the integrators within an org really should just decide when adopting and isn't something that individual plugins should decide. Keeping it as an option lets you set the default in a yarn script. Tbh that point is a lil bit moot tho because we've discussed wanting all wizard decisions to be available both as interactive choices and commandline flags.

**Review by Rugvip — APPROVED** [review]
:shipit:

**Comment by freben** [comment]
I say it's merge ready.

**Comment by benjdlambert** [comment]
So I just had some thoughts whilst migrating some stuff to work with my latest hack - I'm wondering if the `plugin-x-react` package name is semantically correct. I'm thinking that realistically, these plugins aren't really plugins. As they're shared code between other plugins. You can't render and expose them like normal plugins as they're shared components + hooks.  So I was wondering if `@backstage/catalog-react(/web)` is a more apt name for things like this. And then that asks the question should it even live in the plugins folder? Should it be under packages?

**Comment by Fox32** [comment]
Yes, they aren't plugins but packages. I wouldn't put them under packages as there are only core packages. Putting them in a separate directory would help with setting up codeowners. That would also solve problems with packages like catalog-model, that don't match the pattern well.

**Comment by freben** [comment]
We don't want to pollute the `@backstage/x` "space" though - the idea was that by prepending `plugin-` you'd no longer risk future clashes with something that the core team wants to ingroduce. I guess that's a factor/risk worth considering

**Comment by benjdlambert** [comment]
But I mean if they're suffixed with `-react` they shouldn't clash right? Especially when this repo is only going to be the ones contributing to that namespace too I'm having a hard time seeing when that would actually be a problem too?

**Comment by benjdlambert** [comment]
> I wouldn't put them under packages as there are only core packages

I'd argue that the catalog is core anyway? But asides from that, i'm not sure it should be only for 'core packages' as i'm not sure what dictates a core package anyway? It's just something that isn't rendered by `backstage` as a platform directly, and not prefixed with `plugin-` hah :)

**Comment by freben** [comment]
Anything that wouldn't feel weird to re-export from `@backstage/core` could go in `packages` and be "unprefixed" - is that a sensible way of thinking of it?

`catalog-client` is known to be in the wrong place and should go in a future `plugin-catalog-common`. `catalog-model` is arguably confusingly or poorly named, could just as well be `core-entity-model` since it's a shared central integration surface rather than specific to the catalog plugin as such. In the sense that I see it as the core owning/defining that interface and the catalog obeying/storing it, not the other way around

**Comment by benjdlambert** [comment]
But I'm not sure that `plugin-catalog-common` is the right thing to call that? It's not a `plugin` in regards to backstage. It's just an additional library. If we were to create helpers for luxon, we wouldn't call it `plugin-luxon` although I understand the idea of tying it the catalog plugin.

>Anything that wouldn't feel weird to re-export from @backstage/core could go in packages and be "unprefixed" - is that a sensible way of thinking of it?

I thinking not really, I thinking that `packages` are separate node packages that are deployed from our monorepo that are NOT plugins to backstage is what would live in there. Not really tied to `backstage/core` at all to be honest.

**Pull request merged**
